File: src/main/java/org/crafter/engine/gui/components/Button.java
package org.crafter.engine.gui.components;

import org.crafter.engine.camera.Camera;
import org.crafter.engine.gui.enumerators.Alignment;
import org.crafter.engine.gui.factories.FramedMeshFactory;
import org.crafter.engine.gui.font.Font;
import org.crafter.engine.gui.implementations.Text;
import org.crafter.engine.mesh.MeshStorage;
import org.crafter.engine.window.Window;
import org.joml.Vector2f;
import org.joml.Vector2fc;

public class Button extends Text {

    // We want these to be constant throughout the entire game, class members only
    private static final float padding = 16.0f;
    private static final float pixelEdge = 1.0f;
    private static final float borderScale = 2.0f;

    private String buttonBackGroundMeshUUID = null;



    public Button(String name,  String textData, float fontSize, Alignment alignment, Vector2f offset) {
        super(name, textData, fontSize, alignment, offset);
        this._collide = true;

        recalculateMesh();
    }


    @Override
    public void render() {
        Camera.setGuiObjectMatrix(_position.x + getPadding(), _position.y + getPadding());
        MeshStorage.render(this._meshUUID);
        Camera.setGuiObjectMatrix(_position.x, _position.y);
        MeshStorage.render(this.buttonBackGroundMeshUUID);
    }

    @Override
    public boolean collisionDetect(Vector2fc mousePosition) {
        return pointCollisionDetect(mousePosition.x(), mousePosition.y(), _position.x(), _position.y(), _size.x(), _size.y());
    }

    @Override
    protected void recalculateMesh() {
        if (_meshUUID != null) {
            MeshStorage.destroy(_meshUUID);
        }
        if (buttonBackGroundMeshUUID != null) {
            MeshStorage.destroy(buttonBackGroundMeshUUID);
        }

        Vector2f textSize = Font.getTextSize(this.fontSize * getGuiScale(), this.textData);

        buttonBackGroundMeshUUID = FramedMeshFactory.generateMesh(textSize, getPadding(), getPixelEdge(), getBorderScale(), "textures/button.png");

        Font.switchColor(foreGroundColor);
        Font.switchShadowColor(shadowColor);
        _meshUUID = Font.grabText(this.fontSize * getGuiScale(), this.textData);

        // Padding times 2 because all edges of the button are padding, doubled on X and Y
        this.setSize(textSize.add(new Vector2f(getPadding() * 2)));

        this.recalculatePosition();
    }

    @Override
    public void internalOnStep() {
        if (Window.wasResized()) {
            recalculateMesh();
        }
    }

    @Override
    protected void recalculatePosition() {
        this._position.set(_alignment.value().mul(Window.getWindowSize()).sub(getSize().mul(_alignment.value())).add(offset()));
//        System.out.println("Button (" + this.name() + ") POSITION: " + _position.x + ", " + _position.y);
    }

    public static float getPadding() {
        return padding * getGuiScale();
    }

    public static float getPixelEdge() {
        return pixelEdge;
    }

    public static float getBorderScale() {
        return borderScale;
    }

}

File: src/main/kotlin/org/crafter/engine/gui/components/Button.kt
package org.crafter.engine.gui.components

import org.crafter.engine.camera.Camera.setGuiObjectMatrix
import org.crafter.engine.gui.GUI
import org.crafter.engine.gui.enumerators.Alignment
import org.crafter.engine.gui.factories.FramedMeshFactory
import org.crafter.engine.gui.font.Font
import org.crafter.engine.gui.implementations.Text
import org.crafter.engine.mesh.MeshStorage.destroy
import org.crafter.engine.mesh.MeshStorage.render
import org.crafter.engine.window.Window.getWindowSize
import org.crafter.engine.window.Window.wasResized
import org.joml.Vector2f
import org.joml.Vector2fc

class Button(textData: String?, fontSize: Float, alignment: Alignment, offset: Vector2f) :
    Text(textData, fontSize, alignment, offset) {
    private var buttonBackGroundMeshUUID: String = ""

    init {
        collide = true
        recalculateMesh()
    }

    override fun render() {
        setGuiObjectMatrix(position.x + padding, position.y + padding)
        render(meshUUID)
        setGuiObjectMatrix(position.x, position.y)
        render(buttonBackGroundMeshUUID)
    }

    override fun collisionDetect(mousePosition: Vector2fc): Boolean {
        return pointCollisionDetect(
            mousePosition.x(),
            mousePosition.y(),
            position.x(),
            position.y(),
            size.x(),
            size.y()
        )
    }

    override fun recalculateMesh() {
        if (meshUUID != "") {
            destroy(meshUUID)
        }
        if (buttonBackGroundMeshUUID != "") {
            destroy(buttonBackGroundMeshUUID)
        }
        val textSize = Font.getTextSize(fontSize * guiScale, textData)
        buttonBackGroundMeshUUID =
            FramedMeshFactory.generateMesh(textSize, padding, pixelEdge, borderScale, "textures/button.png")
        Font.switchColor(foreGroundColor)
        Font.switchShadowColor(shadowColor)
        meshUUID = Font.grabText(fontSize * guiScale, textData)

        // Padding times 2 because all edges of the button are padding, doubled on X and Y
        this.size = textSize.add(Vector2f(padding * 2))
        recalculatePosition()
    }

    override fun internalOnClick(mousePosition: Vector2fc) {
        TODO("Not yet implemented")
    }

    override fun internalOnStep(gui: GUI) {
        if (wasResized()) {
            recalculateMesh()
        }
    }

    override fun recalculatePosition() {
        position.set(alignment.value().mul(getWindowSize()).sub(size.mul(alignment.value())).add(offset()))
        //        System.out.println("Button (" + this.name() + ") POSITION: " + _position.x + ", " + _position.y);
    }

    companion object {
        // We want these to be constant throughout the entire game, class members only
        val padding = 16.0f
            get() = field * guiScale
        const val pixelEdge = 1.0f
        const val borderScale = 2.0f
    }
}
